planner/core/physical_plan_test.go

```diff
@@ -34,14 +45,150 @@ var _ = Suite(&testPlanSuite{})
 
 type testPlanSuite struct {
 	*parser.Parser
-
 	is infoschema.InfoSchema
+
+	testData testData
+}
+
+// Flag for generate test result.
+var record bool
+
+func init() {
+	flag.BoolVar(&record, "record", false, "to generate test result")
+	flag.Parse()
+}
+
+type testCases struct {
+	Name       string
+	Cases      *json.RawMessage // For delayed parse.
+	decodedOut interface{}      // For generate output.
+}
+
+type testData struct {
+	input          []testCases
+	output         []testCases
+	filePathPrefix string
+	funcMap        map[string]int
 }
 
 func (s *testPlanSuite) SetUpSuite(c *C) {
 	s.is = infoschema.MockInfoSchema([]*model.TableInfo{core.MockSignedTable(), core.MockUnsignedTable()})
 	s.Parser = parser.New()
 	s.Parser.EnableWindowFunc(true)
+
+	var err error
+	s.testData, err = LoadTestSuiteData("testdata", "plan_suite")
+	c.Assert(err, IsNil)
+}
+
+func (s *testPlanSuite) TearDownSuite(c *C) {
+	if record {
+		c.Assert(s.testData.GenerateOutput(), IsNil)
+	}
+}
+
+// LoadTestSuiteData loads test suite data from file.
+func LoadTestSuiteData(dir, suiteName string) (res testData, err error) {
+	res.filePathPrefix = filepath.Join(dir, suiteName)
+	res.input, err = loadTestSuiteCases(fmt.Sprintf("%s_in.json", res.filePathPrefix))
+	if err != nil {
+		return res, err
+	}
+	if record {
+		res.output = make([]testCases, len(res.input), len(res.input))
+		for i := range res.input {
+			res.output[i].Name = res.input[i].Name
+		}
+	} else {
+		res.output, err = loadTestSuiteCases(fmt.Sprintf("%s_out.json", res.filePathPrefix))
+		if err != nil {
+			return res, err
+		}
+		if len(res.input) != len(res.output) {
+			return res, errors.New(fmt.Sprintf("Number of test input cases %d does not match test output cases %d", len(res.input), len(res.output)))
+		}
+	}
+	res.funcMap = make(map[string]int, len(res.input))
+	for i, test := range res.input {
+		res.funcMap[test.Name] = i
+		if test.Name != res.output[i].Name {
+			return res, errors.New(fmt.Sprintf("Input name of the %d-case %s does not match output %s", i, test.Name, res.output[i].Name))
+		}
+	}
+	return res, nil
+}
+
+func loadTestSuiteCases(filePath string) (res []testCases, err error) {
+	jsonFile, err := os.Open(filePath)
+	if err != nil {
+		return res, err
+	}
+	defer jsonFile.Close()
+	byteValue, err := ioutil.ReadAll(jsonFile)
+	if err != nil {
+		return res, err
+	}
+	// Remove comments, since they are not allowed in json.
+	re := regexp.MustCompile("(?s)//.*?\n")
+	err = json.Unmarshal(re.ReplaceAll(byteValue, nil), &res)
+	return res, err
+}
+
+// GetTestCases gets the test cases for a test function.
+func (t *testData) GetTestCases(c *C, in interface{}, out interface{}) {
```

Comment:
Can we move some of them to `util/testutil`?

Reply:
Yes, all of the framework related can be moved to there.